### PR TITLE
python user interface design

### PR DESCRIPTION
**Pull request by Superjom**
some concepts.

resolves: https://github.com/PaddlePaddle/Paddle/issues/3652

### CODE REVIEW

File: doc/design/python/user_interface.md [hunk 1]
```diff
@@ -0,0 +1,362 @@
+# User Interface Design
+
+## Basic Concepts
+### Variable
+A `Variable` represents shared, persistent state manipulated by a Paddle model program.
```
Comment (jacquesqiao) [inline doc/design/python/user_interface.md]: model program?

File: doc/design/python/user_interface.md [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+
+```
+v = pd.Variable(shape=[20,20], initializer=pd.zero_initializer())
+z = pd.Variable(shape=[20,20], initializer=pd.gaussian_initializer(mean=0., std=0.1))
+```
+
+to get the value of the variable, one can call
+
+```python
+print v.val()
```
Comment (jacquesqiao) [inline doc/design/python/user_interface.md]: I think v.eval() is better. because val() means just to get the value, eval() means to calculate and get the value.

File: doc/design/python/user_interface.md [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+# same as 
+v = pd.get_variable(name="v", shape=[20, 20])
+```
+
+By default, Variables are model parameters, and will be updated after the network's back propagation.
+
+One can freeze a variable by setting `trainable` to `False` like:
+
+```python
+v = pd.Variable(shape=[20,20], trainable=False)
```
Comment (jacquesqiao) [inline doc/design/python/user_interface.md]: can also change the state during running.
```python
v = pd.Variable(...)
v.trainable = false
```

File: doc/design/python/user_interface.md (hunk 2 above)
Comment (Superjom) [reply doc/design/python/user_interface.md]: `v.val()` will only return a Python numpy value without executing a sub-graph. `v.eval()` may just be a wrapper of `pd.eval` and will implement it latter if necessary.

File: doc/design/python/user_interface.md (hunk 3 above)
Comment (Superjom) [reply doc/design/python/user_interface.md]: all the arguments in `__init__` are members of `pd.Varaible`, so free to change like that.

File: doc/design/python/user_interface.md (hunk 2 above)
Comment (jacquesqiao) [reply doc/design/python/user_interface.md]: ok

File: doc/design/python/user_interface.md [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+```
+which means that a operator `pd.matmul` takes two variables `a` and `b` for input, 
+and return a variable `out`.
+
+### Layer
+`Layer` defines a more complex operation which may combines several `Op`s, its usage is the same with `Op`.
+
+Take `pd.fc` for example, one can use it like this
+```python
+out = pd.fc(in, param_names=['W'])
```
Comment (jacquesqiao) [inline doc/design/python/user_interface.md]: maybe 
```python
out = pd.fc(int, W="w", B="b")
```

File: doc/design/python/user_interface.md [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+#### Initializer Ops
+These ops will initialize variables, for example, we may have 
+
+- `pd.zero_initializer()`
+- `pd.gaussian_random_initializer(mean, std)`
+
+Each trainable variable has a initialize Op. 
+
+#### Optimizer Ops
+These ops will help to optimize trainable variables after backward propagation finished,
```
Comment (jacquesqiao) [inline doc/design/python/user_interface.md]: backpropagation？

File: doc/design/python/user_interface.md [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+
+v = pd.Variable(shape=[20, 20])
+```
+
+To make it more converient to share a variable, each `pd.Variable` has a name, 
+one can use a name to get or create a `pd.Variable` by calling `pd.get_variable`, for example:
+
+```python
+# same as 
+v = pd.get_variable(name="v", shape=[20, 20])
```
Comment (reyoung) [inline doc/design/python/user_interface.md]: get_varibale from what scope?

File: python/paddle/python_wrapper_demo/block.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    '''
+    Block is the concept of code block, which has a sequence of local Variables
+    and Operators.
+    '''
+
+    def __init__(self):
+        '''
+        namespace: str
+        '''
+        self.cmds = []
```
Comment (reyoung) [inline python/paddle/python_wrapper_demo/block.py]: Maybe ops is better than cmds.

File: python/paddle/python_wrapper_demo/block.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        ops = self.__extract_op_from_block(self.cmds)
+        for op in ops:
+            self.net.append_op(op)
+        self.net.complete_add_op(True)
+
+    def __extract_op_from_block(self, cmds):
+        ops = []
+        for cmd in cmds:
+            if type(cmd) is Block:
+                child_ops = self.__extract_op_from_block([cmd])
```
Comment (reyoung) [inline python/paddle/python_wrapper_demo/block.py]: Why we must extract op from block?

File: python/paddle/python_wrapper_demo/variable.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    '''
+    counter = 0
+    __varset__ = set()
+
+    def __init__(self,
+                 name=None,
+                 shape=[],
+                 data=None,
+                 initialzier=None,
+                 scope=g_scope,
```
Comment (reyoung) [inline python/paddle/python_wrapper_demo/variable.py]: In python, complex instance cannot be default value.

File: doc/design/python/user_interface.md [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+    pd.cross_entropy(data=D_real_fake, label=zeros_label))
+D_loss = D_loss_real + D_loss_fake
+
+G_loss = pd.reduce_mean(pd.cross_entropy(data=D_loss_fake, label=ones_label))
+
+D_solver = pd.AdamOptimizer().minimize(D_loss, var_list=theta_D)
+G_solver = pd.AdamOptimizer().minimize(G_loss, var_list=theta_G)
+
+# init all parameters
+initializer = pd.variable_initialzier()
```
Comment (reyoung) [inline doc/design/python/user_interface.md]: variable_initialziers

File: python/paddle/python_wrapper_demo/namespace.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+
+class Namespace(object):
+    '''
+    Namespace is similar to tf.variable_scope, which helps to store local
+    variables  of different `Block`s into one Scope, and make it possible
+    to reference Variable across different Blocks.
+    '''
+
+    stack = []
```
Comment (reyoung) [inline python/paddle/python_wrapper_demo/namespace.py]: How could a user want two global namespaces? `stack` is a data member in class scope. So all instances of `Namespace` share the same `stack`.

File: python/paddle/python_wrapper_demo/block.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+        Namespace.begin(self.namespace)
+        block_guard.last_block = block_guard.cur_block
+        block_guard.cur_block = self.block
+
+        if self.execute_immediately:
+            block_guard.last_block.append(block_guard.cur_block)
+
+    def __exit__(self):
+        Namespace.end()
+        block_guard.cur_block = block_guard.last_block
```
Comment (reyoung) [inline python/paddle/python_wrapper_demo/block.py]: Here the current block is dropped in Python. Which object could store that block? Should that block be freed?

File: python/paddle/python_wrapper_demo/variable.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class Variable(object):
+    '''
+    Variable is the data type of Operator's inputs and outputs.
+    '''
+    counter = 0
+    __varset__ = set()
+
+    def __init__(self,
```
Comment (reyoung) [inline python/paddle/python_wrapper_demo/variable.py]: 正如我们之前电话沟通过的，只通过名字，是无法定位运行到神经网络的哪个子图的。因为在一个神经网络中，可以有两个Op输出同名的Variable。

Comment (jacquesqiao) [reply python/paddle/python_wrapper_demo/variable.py]: 这个问题，后来想了一下，貌似终点是op也可能碰到同样的问题？ 比如以`可以有两个Op输出同名的Variable` 这个Variable作为input的Op

File: doc/design/python/user_interface.md [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+```
+which means that a operator `pd.matmul` takes two variables `a` and `b` for input, 
+and return a variable `out`.
+
+### Layer
+`Layer` defines a more complex operation which may combines several `Op`s, its usage is the same with `Op`.
+
+Take `pd.fc` for example, one can use it like this
+```python
+out = pd.fc(in, param_names=['W'])
```
Comment (Canpio) [inline doc/design/python/user_interface.md]: 1. Is parameter `w` defined outside of fc layer or inside the layer?
2. If a layer has more than one parameter, how can we know the name correspondence?

Comment (Superjom) [reply doc/design/python/user_interface.md]: currently, maybe layer create parameters better. the big difference between a op and a layer.

File: doc/design/python/user_interface.md [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+
+```python
+import paddle as pd
+
+# the first shape is None, which means the batch size of variable is not known.
+image = pd.Variable(shape=[None, 128])
+label = pd.Variable(shape=[None, 1])
+
+# network config
+W1 = pd.Variable('W1', shape=[128, 64])
```
Comment (Canpio) [inline doc/design/python/user_interface.md]: name = 'W1' ?

File: python/paddle/python_wrapper_demo/block.py (hunk 7 above)
Comment (Superjom) [reply python/paddle/python_wrapper_demo/block.py]: block may contain block ... 

```
{// block 
   {// sub-block
   }
}
```

File: python/paddle/python_wrapper_demo/block.py (hunk 12 above)
Comment (Superjom) [reply python/paddle/python_wrapper_demo/block.py]: sub-block will inserted into its father-block, so it will not be freed

```c++
// father-scope [ op1, op2, sub-block, op3, op4]

{// father-scope
   {// sub-scope
   }
}
```

File: python/paddle/python_wrapper_demo/namespace.py (hunk 11 above)
Comment (Superjom) [reply python/paddle/python_wrapper_demo/namespace.py]: yes, there needs only one namespace.stack, because user's config is parsed using only one thread.

File: python/paddle/python_wrapper_demo/variable.py (hunk 13 above)
Comment (Superjom) [reply python/paddle/python_wrapper_demo/variable.py]: here, each variable will have a unique name across all the sub-graphs, using `"var-%d" % (Var.couter++)"`

A specific namespace will add some prefix to a Variable's name, and help to support local variables.

File: doc/design/python/user_interface.md (hunk 6 above)
Comment (Superjom) [reply doc/design/python/user_interface.md]: get_variable from current namespace, in user's model config, they need only one global scope, but different namespace(add prefix to var's name).

The difference between multiple scope and namespace is that scopes will turn into a forest, different sub-scopes without the same ancestor is totally separated, but namespace doesn't suffer from this:

- all the variable are located in a global scope
- different namespaces have a different name prefix

File: doc/design/python/user_interface.md (hunk 10 above)
Comment (Superjom) [reply doc/design/python/user_interface.md]: yes

File: python/paddle/python_wrapper_demo/block.py (hunk 8 above)
Comment (Superjom) [reply python/paddle/python_wrapper_demo/block.py]: Currently, when a block executes, it will create a NetOp and run it, so operators from multiple block should be extracted and insert into a NetOp. 

This may be changed later to make it more natural.

File: python/paddle/python_wrapper_demo/variable.py (hunk 9 above)
Comment (Superjom) [reply python/paddle/python_wrapper_demo/variable.py]:
```python
In [1]: class A(object):
   ...:     def __init__(self):
   ...:         self.x = None
   ...:         

In [2]: def f(x=A()):
   ...:     print x
   ...:     

In [3]: f()
<__main__.A object at 0x10c7cc050>
```

this seems works

**Comment by Superjom** [comment]
@reyoung 

1. I think layer can infer the parameters' shape easily, but an op can only have input, a parameter is treated as a normal input, it seems natural that an op can infer its output, but weird to infer its inputs.
2. V2 is treated as a higher interface, which is like Keras to TF, and this design is the underlying API which is similar to TF. It is right that MNist's demo is exactly the same as the V2's.  
3. Will add this latter, TODO
4. In the implementation, two functions are implemented, `__enter__` and `__exit__`, in another language, if it has something similar to `with` statement semantically, it can support a similar 
syntax.
5. they are exactly the same, just one thing, use `is_trainable` and delete `is_param`.  By default, `is_trainable` is true, if user or Layer create a Variable, that should be treated as a parameter, and all other temp outputs of op/layer are not `trainable`.
6. They are the same role, may remove RNNOp and add some functions to Block, I will add a design doc about Block.
7. Only `Op`, `Variable`, `namespace` are new, all others will be hidden in the underlying implementation.

File: python/paddle/python_wrapper_demo/namespace.py (hunk 11 above)
Comment (reyoung) [reply python/paddle/python_wrapper_demo/namespace.py]: 不是一个线程，而是一个进程。

在这个设计里，用户似乎无法在同一个进程中配置多个神经网络。

File: python/paddle/python_wrapper_demo/variable.py (hunk 9 above)
Comment (reyoung) [reply python/paddle/python_wrapper_demo/variable.py]: See http://effbot.org/zone/default-values.htm

**Comment by reyoung** [comment]
@Superjom 
I think following comments are not responsed.

> How can the user configure two neural networks independently? Because the with modify the same global instance, it is no other way that user does not use that instance. If I misunderstood this, please give a sample code for the following situation. The user uses mnist dataset, and configure a convolutional network and a fast forward network in same program. He needs to know which network is better. So he runs same mini-batch both on two networks step by step.

> We must inference the shape of operator outputs line by line. Every time we create an operator, we must know the output shape of that operator immediately.


另外

> Only Op, Variable, namespace are new, all others will be hidden in the underlying implementation.

隐藏的复杂性也是复杂性。真的有必要引入这么多新的概念么？可不可以用更少的概念实现这个设计？

譬如，namespace看起来和guard是配合使用的，也许可以合并起来。当然，我并不完全能够掌握这个设计，可能有些吹毛求疵了

**Comment by Superjom** [comment]
@reyoung 

a minist with multiple sub-models:

```python
import paddle as pd

image = pd.Variable([None, 128])
label = pd.Variable([None, 1])

def FC_model():
    # every Varialbe will get a unique name like "var-%d"%(Variable.counter++)
    # both layer's parameters and temp outputs are stored in a global scope with unique names
    # so it is OK to configure a submodel in a python function
    # and any variable can be passed as argument across all the python scopes.
    fc_out = pd.fc(image)
    pred = pd.softmax(fc_out, size=10)
    return pred

def CNN_model():
    out = pd.conv_group(image, xxx)
    pred = pd.softmax(out, size=10)
    return pred

def data_reader(path):
    xxxx
    yield batch

def run_model(pred, batch):
    cost = pd.cross_entropy(pred, label)
    optimizer = pd.SGDOptimizer().minimize([cost])
    _, cost_v = pd.eval([optimizer, cost], feeds={image:batch[0], label:batch[1]})
    return cost_v

data_provider = data_reader('./data.txt')
a_batch_of_data = data_provider.next()

print 'fc_cost', run_model(FC_model(), a_batch_of_data)
print 'cnn_cost', run_model(CNN_model(), a_batch_of_data)
```

About new concepts:

- Only the Block is a hidden new concept now?  but NetOp will be replaced to Block, so new concepts is not too much, and in the user interface, `pd.Block` is hidden to users, that's why I remove `pd.Block` in the design doc. 
- Currently, Block seems a little complex, behaves differently in compile and execution period, and I will add a new design doc about Block latter, and we can have more discussion about that.

**Comment by Superjom** [comment]
merge namespace with its guard, good idea.

I merge Block with its guard, but forget to merge the namespace's, will update the code latter.
@reyoung
